File: day02/part1.py
from common import main
from day02.input import parse_input


def find_invalid_ids(start: int, end: int) -> list[int]:
    start_str = str(start)

    if len(start_str) & 1 == 0:
        candidate_half = int(start_str[:len(start_str) // 2])
    else:
        # Just an optimization to cut out the odd lengths
        candidate_half = 10 ** (len(start_str) // 2)
    invalid_ids = []

    candidate = int(str(candidate_half) + str(candidate_half))
    while candidate <= end:
        if candidate >= start:
            invalid_ids.append(candidate)

        candidate_half += 1
        candidate = int(str(candidate_half) + str(candidate_half))

    return invalid_ids

def solve_part1(id_ranges: list[tuple[int, int]]):
    invalid_ids = [
        invalid_id
        for range in id_ranges
        for invalid_id in find_invalid_ids(range[0], range[1])
    ]

    return sum(invalid_ids)

if __name__ == "__main__":
    main(parse_input, solve_part1)

File: day02/part1.rs
mod input;

use common::run;
use input::{parse_input, Input};

fn find_invalid_ids(start: usize, end: usize) -> Vec<usize> {
    let start_str = start.to_string();

    let mut candidate_half = if start_str.len() & 1 == 0 {
        start_str[..start_str.len() / 2].parse::<usize>().unwrap()
    } else {
        // Just an optimization to cut out the odd lengths
        10_usize.pow((start_str.len() / 2) as u32)
    };

    let mut invalid_ids = Vec::new();

    let mut candidate = format!("{}{}", candidate_half, candidate_half).parse::<usize>().unwrap();
    while candidate <= end {
        if candidate >= start {
            invalid_ids.push(candidate);
        }

        candidate_half += 1;
        candidate = format!("{}{}", candidate_half, candidate_half).parse::<usize>().unwrap();
    }

    invalid_ids
}

fn solve_part1(input: Input) -> i64 {
	input.into_iter()
		.flat_map(|(start, end)| find_invalid_ids(start, end))
		.sum::<usize>() as i64
}

fn main() {
	run(parse_input, solve_part1);
}
